internal/server/server.go

```diff
@@ -108,19 +198,92 @@ func (s *Server) handleLogin() http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		session := s.getSession(r)
 
-		rs, err := random.SecureString(32)
+		randomStr, err := random.SecureString(32)
+		if err != nil {
+			logAndError(w, http.StatusInternalServerError, err, "error generating random string for state")
+			return
+		}
+
+		// generate nonce
+		nonce, err := random.SecureString(32)
 		if err != nil {
-			logAndError(w, http.StatusInternalServerError, err, "error generating random string")
+			logAndError(w, http.StatusInternalServerError, err, "error generating random string for nonce")
 			return
 		}
 
-		session.Values["rand"] = rs
+		session.Values["nonce"] = nonce
 
 		if err := session.Save(r, w); err != nil {
 			logAndError(w, http.StatusInternalServerError, err, "error saving session")
 			return
 		}
 
-		fmt.Fprintf(w, "session saved")
+		http.Redirect(w, r, s.oauth2Config.AuthCodeURL(randomStr, oidc.Nonce(nonce)), http.StatusSeeOther)
+	}
+}
+
+func (s *Server) handleCallback() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// handle requests that contain an error
+		if err := r.URL.Query().Get("error"); err != "" {
+			logAndError(w, http.StatusBadRequest, fmt.Errorf(err+": "+r.URL.Query().Get("error_description")), "error in request")
+			return
+		}
+
+		session := s.getSession(r)
+
+		// convert authorization code into an OAuth2 token
+		oauth2Token, err := s.externalFuncs.oauth2ConfigExchange(s.oauth2Config, s.context, r.URL.Query().Get("code"))
```

Comment:
Since this is an incoming request handler it should instead use the passed in request's context.

https://pkg.go.dev/net/http#Request.Context

```suggestion
		oidcContext := oidc.ClientContext(r.Context(), s.client)
		oauth2Token, err := s.externalFuncs.oauth2ConfigExchange(s.oauth2Config, oidcContext, r.URL.Query().Get("code"))
```

Reply:
Fixed